optuna/visualization/_pareto_front.py

```diff
@@ -171,53 +130,200 @@ def _make_scatter_object(
+        for trial in study.get_trials(states=(TrialState.COMPLETE,)):
+            if all(map(lambda x: x <= 0.0, constraints_func(trial))):
+                feasible_trials.append(trial)
+            else:
+                infeasible_trials.append(trial)
+        best_trials = _get_pareto_front_trials_by_trials(feasible_trials, study.directions)
+        if include_dominated_trials:
+            non_best_trials = _get_non_pareto_front_trials(feasible_trials, best_trials)
+        else:
+            non_best_trials = []
+
+        if len(best_trials) == 0:
+            _logger.warning("Your study does not have any completed and feasible trials.")
+    else:
+        best_trials = study.best_trials
+        if len(best_trials) == 0:
+            _logger.warning("Your study does not have any completed trials.")
+
+        if include_dominated_trials:
+            non_best_trials = _get_non_pareto_front_trials(
+                study.get_trials(deepcopy=False), best_trials
+            )
+        else:
+            non_best_trials = []
+        infeasible_trials = []
+
+    _targets = targets
+    if _targets is None:
+        if len(study.directions) == 2:
+            _targets = _targets_default_2d
+        elif len(study.directions) == 3:
+            _targets = _targets_default_3d
+        else:
+            raise ValueError(
+                "`plot_pareto_front` function only supports 2 or 3 objective"
+                " studies when using `targets` is `None`. Please use `targets`"
+                " if your objective studies have more than 3 objectives."
+            )
+
+    def _make_trials_with_values(
+        trials: Optional[List[FrozenTrial]],
+        targets: Callable[[FrozenTrial], Sequence[float]],
+    ) -> Optional[Sequence[Tuple[FrozenTrial, Sequence[float]]]]:
+        if trials is None:
+            return None
+        return [(trial, targets(trial)) for trial in trials]
+
+    best_trials_with_values = _make_trials_with_values(best_trials, _targets)
+    non_best_trials_with_values = _make_trials_with_values(non_best_trials, _targets)
+    infeasible_trials_with_values = _make_trials_with_values(infeasible_trials, _targets)
+
+    def _infer_n_targets(
+        trials_with_values: Optional[Sequence[Tuple[FrozenTrial, Sequence[float]]]]
+    ) -> Optional[int]:
+        if trials_with_values is not None and len(trials_with_values) > 0:
+            if not isinstance(trials_with_values[0][1], collections.abc.Sequence):
+                raise ValueError(
+                    "`targets` should return a sequence of target values."
+                    " your `targets` returns {}".format(type(trials_with_values[0][1]))
+                )
+            return len(trials_with_values[0][1])
+        return None
+
+    n_targets = (
+        _infer_n_targets(best_trials_with_values)
+        or _infer_n_targets(non_best_trials_with_values)
+        or _infer_n_targets(infeasible_trials_with_values)
+    )
```

Comment:
[question]

Do we have to invoke `_infer_n_targets` for `non_best_trials_with_values` here? I expect `_infer_n_targets` is invoked for each type of trials because `best_trials_with_values` would be `None` when no feasible trial could be available. 

But, I think `non_best_trials_with_values` would be also empty when `best_trials_with_values` is empty, and so it'd be enough to check `best_trials_with_values` and `infeasible_trials_with_values`. Please correct me if I don't understand your intention.

Reply:
I was unaware that we can skip check for `non_best_trials_with_values`. I think just removing the check may be a little confusing, so I will leave a short comment here.